src/jbuf/jbuf.c

```diff
@@ -295,67 +348,80 @@ int jbuf_set_type(struct jbuf *jb, enum jbuf_type jbtype)
 }
 
 
-static void wish_down(void *arg)
+/**
+ * Moves end of complete sequence and counts complete frames. If `end` is NULL,
+ * a search starts from the head which points to the oldest packet
+ *
+ * @param jb   Jitter buffer
+ * @param cur  Current inserted list element
+ */
+static void jbuf_move_end(struct jbuf *jb, struct le *cur)
 {
-	struct jbuf *jb = arg;
+	struct le *end;
+
+	if (!jb->end) {
+		jb->ncf  = 0;
+		cur = jb->packetl.head;
+		if (!cur)
+			return;
+
+		jb->end = cur;
+		cur = cur->next;
+		if (!cur)
+			return;
+	}
+
+	end = cur->prev;
+	if (!end)
+		return;
+
+	/* update only if endsing packet was inserted right now */
+	if (jb->end != end)
+		return;
+
+	for (; end->next; end = end->next) {
+		struct packet *pm = end->data;
+		struct packet *pn = end->next->data;
 
-	if (jb->wish > jb->min) {
-		DEBUG_INFO("wish size changed %u --> %u\n", jb->wish,
-			   jb->wish - 1);
-		--jb->wish;
+		if (pm->hdr.seq + 1 != pn->hdr.seq)
+			break;
+
+		if (pm->hdr.ts != pn->hdr.ts)
+			++jb->ncf;
 	}
+
+	jb->end = end;
 }
 
 
-static void calc_rdiff(struct jbuf *jb, uint16_t seq)
+static bool jbuf_frame_ready(struct jbuf *jb)
 {
-	int32_t rdiff;
-	int32_t adiff;
-	int32_t s;                         /**< EMA coefficient              */
-	float ratio = 1.0;                 /**< Frame packet ratio           */
-	uint32_t wish;
-	uint32_t max = jb->max;
-	bool down = false;
-
-	if (jb->jbtype != JBUF_ADAPTIVE)
-		return;
+	if (!jb->packetl.head)
+		return false;
 
-	if (!jb->seq_get)
-		return;
+	if (jb->nf < jb->min)
+		return false;
 
-	if (jb->nf) {
-		ratio = (float)jb->n / (float)jb->nf;
-		max   = (uint32_t)(max / ratio);
-	}
+	return jb->ncf || jb->nf > jb->max;
+}
 
-	rdiff = (int16_t)(jb->seq_put + 1 - seq);
-	adiff = abs(rdiff * JBUF_RDIFF_EMA_COEFF);
-	s = adiff > jb->rdiff ? JBUF_RDIFF_UP_SPEED :
-		jb->wish > 2  ? 1 :
-		jb->wish > 1  ? 2 : 3;
-	jb->rdiff += (adiff - jb->rdiff) * s / JBUF_RDIFF_EMA_COEFF;
 
-	wish = (uint32_t)(jb->rdiff / (float)JBUF_RDIFF_EMA_COEFF / ratio);
-	if (wish < jb->min)
-		wish = jb->min;
+static void reset_wait(void *arg)
+{
+	struct jbuf *jb = arg;
 
-	if (max && wish >= max)
-		wish = max - 1;
+	jb->wait  = false;
+	jb->again = true;
+}
 
-	if (wish > jb->wish) {
-		DEBUG_INFO("wish size changed %u --> %u\n", jb->wish, wish);
-		jb->wish = wish;
-	}
-	else if (wish < jb->wish) {
-		uint32_t dt = wish + 1 == jb->wish ? 6000 : 1000;
-		if (!tmr_isrunning(&jb->tmr) || tmr_get_expire(&jb->tmr) > dt)
-			tmr_start(&jb->tmr, dt, wish_down, jb);
 
-		down = true;
-	}
+static void eagain_later(struct jbuf *jb)
+{
+	jb->wait = true;
+	if (tmr_isrunning(&jb->tmr))
+		return;
 
-	if (!down && tmr_isrunning(&jb->tmr))
-		tmr_cancel(&jb->tmr);
+	tmr_start(&jb->tmr, 250, reset_wait, jb);
```

Comment:
Does this fixed `250ms` value works well with different `fps` and `min/max` jbuf combinations?

Reply:
`rdiff` didn't solve problem for audio. It was only 2 packets while number of complete frames need to be higher to avoid underruns in following aubuf.